win32/devdoc/file_win32_requirements.md

```diff
@@ -0,0 +1,178 @@
+
+**SRS_FILE_WIN32_43_002: [** `file_create` shall call `SetFileCompletionNotificationModes` to disable calling the completion port when an async operations finishes synchrounously. **]**
+
+**SRS_FILE_WIN32_43_003: [** `file_create` shall allocate a threadpool environment. **]**
+
+**SRS_FILE_WIN32_43_004: [** `file_create` shall obtain a `PTP_POOL` struct by calling `execution_engine_win32_get_threadpool` on `execution_engine`. **]**
+
+**SRS_FILE_WIN32_43_005: [** `file_create` shall register the threadpool environment with `ptpPool` **]**
+
+**SRS_FILE_WIN32_43_006: [** `file_create` shall create a cleanup group. **]**
+
+**SRS_FILE_WIN32_43_007: [** `file_create` shall register the cleanup group with the threadpool environment. **]**
+
+**SRS_FILE_WIN32_43_033: [** `file_create` shall create a threadpool io with the allocated `FILE_HANDLE` and `on_file_io_complete_win32` as a callback. **]**
+
+**SRS_FILE_WIN32_43_008: [** If there are any failures, `file_create` shall return `NULL`. **]**
+
+**SRS_FILE_WIN32_43_009: [** `file_create` shall succeed and return a non-`NULL` value. **]**
+
+## file_destroy
+
+```c
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+```
+
+**SRS_FILE_WIN32_43_049: [** If `handle` is `NULL`, `file_destroy` shall return. **]**
+
+**SRS_FILE_WIN32_43_011: [** `file_destroy` shall wait for all I/O to complete. **]**
+
+**SRS_FILE_WIN32_43_012: [** `file_destroy` shall close the cleanup group. **]**
+
+**SRS_FILE_WIN32_43_013: [** `file_destroy` shall destroy the environment. **]**
+
+**SRS_FILE_WIN32_43_016: [** `file_destroy` shall call `CloseHandle` on `handle` **]**
+
+**SRS_FILE_WIN32_43_015: [** `file_destroy` shall close the threadpool IO. **]**
+
+**SRS_FILE_WIN32_43_042: [** `file_destroy` shall free the `FILE_HANDLE`. **]**
+
+
+## file_write_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(FILE_WRITE_ASYNC_OK, FILE_WRITE_ASYNC_ERROR);
+```
+
+**SRS_FILE_WIN32_43_043: [** If `handle` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_WIN32_43_044: [** If `source` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_WIN32_43_045: [** If `user_callback` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_WIN32_43_017: [** `file_write_async` shall call `StartThreadpoolIo`. **]**
+
+**SRS_FILE_WIN32_43_020: [** `file_write_async` shall allocate an `OVERLAPPED` struct and populate it with an event and `position`. **]**

```

Comment:
>event [](start = 110, length = 5)

event needs to created by CreateEvent first, so I guess that's a different spec.

Reply:
Added spec in next iteration.

---
In reply to: [452473165](https://github.com/Azure/azure-c-pal/pull/7#discussion_r452473165) [](ancestors = 452473165)